File: dautari_adda/lib/features/admin/data/roles_service.dart
import 'dart:convert';
import 'package:dautari_adda/core/api/api_service.dart';

class RolesService {
  final ApiService _apiService = ApiService();

  // Get all roles
  Future<List<dynamic>> getRoles({
    int skip = 0,
    int limit = 100,
  }) async {
    try {
      final response = await _apiService.get('/roles?skip=$skip&limit=$limit');
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      }
      return [];
    } catch (e) {
      return [];
    }
  }

  // Get specific role
  Future<Map<String, dynamic>?> getRole(int roleId) async {
    try {
      final response = await _apiService.get('/roles/$roleId');
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      }
      return null;
    } catch (e) {
      return null;
    }
  }

  // Get available permissions
  Future<List<String>> getAvailablePermissions() async {
    try {
      final response = await _apiService.get('/roles/permissions');
      if (response.statusCode == 200) {
        return List<String>.from(jsonDecode(response.body));
      }
      return [];
    } catch (e) {
      return [];
    }
  }

  // Create new role (Admin only)
  Future<Map<String, dynamic>> createRole({
    required String name,
    String? description,
    List<String>? permissions,
  }) async {
    try {
      final response = await _apiService.post('/roles', {
        'name': name,
        if (description != null) 'description': description,
        if (permissions != null) 'permissions': permissions,
      });

      final data = jsonDecode(response.body);
      if (response.statusCode == 201) {
        return {'success': true, 'data': data};
      } else {
        throw data['detail'] ?? 'Failed to create role';
      }
    } catch (e) {
      if (e is String) rethrow;
      throw 'Failed to create role: $e';
    }
  }

  // Update role (Admin only)
  Future<bool> updateRole(int roleId, Map<String, dynamic> data) async {
    try {
      final response = await _apiService.put('/roles/$roleId', data);
      return response.statusCode == 200;
    } catch (e) {
      return false;
    }
  }

  // Delete role (Admin only)
  Future<bool> deleteRole(int roleId) async {
    try {
      final response = await _apiService.delete('/roles/$roleId');
      return response.statusCode == 200;
    } catch (e) {
      return false;
    }
  }
}

File: Ratala-POS-System-Web/backend/app/services/roles_service.py
"""
Role service for managing dynamic roles and permissions
"""
from sqlalchemy.orm import Session
from typing import List, Optional
from app.models.role import Role
from app.schemas import RoleCreate, RoleUpdate


def get_roles(db: Session, skip: int = 0, limit: int = 100) -> List[Role]:
    """Get all roles"""
    return db.query(Role).offset(skip).limit(limit).all()


def get_role(db: Session, role_id: int) -> Optional[Role]:
    """Get a single role by ID"""
    return db.query(Role).filter(Role.id == role_id).first()


def get_role_by_name(db: Session, name: str) -> Optional[Role]:
    """Get a single role by name"""
    return db.query(Role).filter(Role.name == name).first()


def create_role(db: Session, role_data: RoleCreate) -> Role:
    """Create a new role"""
    db_role = Role(
        name=role_data.name,
        description=role_data.description,
        permissions=role_data.permissions
    )
    db.add(db_role)
    db.commit()
    db.refresh(db_role)
    return db_role


def update_role(db: Session, role_id: int, role_data: RoleUpdate) -> Optional[Role]:
    """Update an existing role"""
    db_role = get_role(db, role_id)
    if not db_role:
        return None
    
    update_data = role_data.model_dump(exclude_unset=True)
    for key, value in update_data.items():
        setattr(db_role, key, value)
    
    db.commit()
    db.refresh(db_role)
    return db_role


def delete_role(db: Session, role_id: int) -> bool:
    """Delete a role"""
    db_role = get_role(db, role_id)
    if not db_role:
        return False
    
    db.delete(db_role)
    db.commit()
    return True


def get_available_permissions() -> List[str]:
    """
    Get all available granular permissions in the system.
    This can be expanded as more features are added.
    """
    return [
        "dashboard.view",
        "pos.access"
    ]
